src/main/java/com/yahoo/bullet/bql/query/ProcessedQuery.java

```diff
@@ -0,0 +1,360 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.yahoo.bullet.querying.aggregations.grouping.GroupOperation.GroupOperationType.COUNT;
+
+@Getter
+public class ProcessedQuery {
+    public enum QueryType {
+        TOP_K,
+        DISTRIBUTION,
+        COUNT_DISTINCT,
+        GROUP,
+        SELECT_DISTINCT,
+        SELECT_ALL,
+        SELECT,
+        SPECIAL_K,
+        INVALID
+    }
+
+    private Set<QueryType> queryTypeSet = EnumSet.noneOf(QueryType.class);
+
+    @Setter
+    private Long timeDuration;
+    @Setter
+    private Integer limit;
+    @Setter
+    private WindowNode window;
+    @Setter
+    private ExpressionNode whereNode;
+    @Setter
+    private ExpressionNode havingNode;
+
+    private Map<ExpressionNode, Expression> preAggregationMapping = new HashMap<>();
+    private Map<ExpressionNode, Expression> postAggregationMapping = new HashMap<>();
+
+    private Map<ExpressionNode, String> aliases = new HashMap<>();
+    private Set<ExpressionNode> subExpressionNodes = new HashSet<>();
+
+    private List<SelectItemNode> selectNodes = new ArrayList<>();
+    private List<ExpressionNode> groupByNodes = new ArrayList<>();
+    private List<SortItemNode> orderByNodes = new ArrayList<>();
+
+    private Set<ExpressionNode> superAggregateNodes = new HashSet<>();
+    private Set<ExpressionNode> aggregateNodes = new HashSet<>();
+    private Set<GroupOperationNode> groupOpNodes = new HashSet<>();
+    private Set<CountDistinctNode> countDistinctNodes = new HashSet<>();
+    private Set<DistributionNode> distributionNodes = new HashSet<>();
+    private Set<TopKNode> topKNodes = new HashSet<>();
+
+    private List<BulletError> errors = new ArrayList<>();
+    @Setter
+    private Collection<ExpressionNode> projection;
+    @Setter
+    private Collection<ExpressionNode> computation;
+
+    /**
+     * Validates the query components.
+     *
+     * @return This {@link ProcessedQuery} for convenience.
+     */
+    public ProcessedQuery validate() {
+        if (queryTypeSet.size() != 1) {
+            errors.add(new BulletError("Query does not match exactly one query type: " + queryTypeSet, "Please specify a valid query."));
+        }
+        if (aliases.values().contains("")) {
+            errors.add(new BulletError("Cannot have an empty string as an alias.", "Please specify a non-empty string as the alias."));
+        }
+        if (countDistinctNodes.size() > 1) {
+            errors.add(new BulletError("Cannot have multiple count distincts.", "Please specify only one count distinct."));
+        }
+        if (distributionNodes.size() > 1) {
+            errors.add(new BulletError("Cannot have multiple distributions.", "Please specify only one distribution."));
+        }
+        if (topKNodes.size() > 1) {
+            errors.add(new BulletError("Cannot have multiple top k.", "Please specify only one top k."));
```

Comment:
Well, personally, I want to keep things not in all caps, and we can change those in the appropriate places in bullet-core. Just because there are enums in the code doesn't mean that errors (which should be legible) should have ALL CAPS.

Reply:
I don't mind either way. Just want error messages that go back to the user to have a consistent style.